pkg/generators/merge.go

```diff
@@ -0,0 +1,201 @@
+	for _, generator := range generators {
+		generatorParamSets, err := m.getParams(generator, appSet)
+		if err != nil {
+			return nil, err
+		}
+		// concatenate param lists produced by each generator
+		paramSets = append(paramSets, generatorParamSets)
+	}
+	return paramSets, nil
+}
+
+// GenerateParams gets the params produced by the MergeGenerator.
+func (m *MergeGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.ApplicationSetGenerator, appSet *argoprojiov1alpha1.ApplicationSet) ([]map[string]string, error) {
+	if appSetGenerator.Merge == nil {
+		return nil, EmptyAppSetGeneratorError
+	}
+
+	if len(appSetGenerator.Merge.Generators) < 2 {
+		return nil, LessThanTwoGeneratorsInMerge
+	}
+
+	paramSetsFromGenerators, err := m.getParamSetsForAllGenerators(appSetGenerator.Merge.Generators, appSet)
+	if err != nil {
+		return nil, err
+	}
+
+	baseParamSetsByMergeKey, err := getParamSetsByMergeKey(appSetGenerator.Merge.MergeKeys, paramSetsFromGenerators[0])
+	if err != nil {
+		return nil, err
+	}
+
+	for _, paramSets := range paramSetsFromGenerators[1:] {
+		paramSetsByMergeKey, err := getParamSetsByMergeKey(appSetGenerator.Merge.MergeKeys, paramSets)
+		if err != nil {
+			return nil, err
+		}
+
+		for mergeKeyValue, baseParamSet := range baseParamSetsByMergeKey {
+			if overrideParamSet, exists := paramSetsByMergeKey[mergeKeyValue]; exists {
+				overriddenParamSet, err := utils.CombineStringMapsAllowDuplicates(baseParamSet, overrideParamSet)
+				if err != nil {
+					return nil, err
+				}
+				baseParamSetsByMergeKey[mergeKeyValue] = overriddenParamSet
+			}
+		}
+	}
+
+	mergedParamSets := make([]map[string]string, len(baseParamSetsByMergeKey))
+	var i = 0
+	for _, mergedParamSet := range baseParamSetsByMergeKey {
+		mergedParamSets[i] = mergedParamSet
+		i += 1
+	}
+
+	return mergedParamSets, nil
+}
+
+// getParamSetsByMergeKey converts the given list of parameter sets to a map of parameter sets where the key is the
+// unique key of the parameter set as determined by the given mergeKeys. If any two parameter sets share the same merge
+// key, getParamSetsByMergeKey will throw NonUniqueParamSets.
+func getParamSetsByMergeKey(mergeKeys []string, paramSets []map[string]string) (map[string]map[string]string, error) {
+	if len(mergeKeys) < 1 {
+		return nil, NoMergeKeys
+	}
+
+	deDuplicatedMergeKeys := make(map[string]bool, len(mergeKeys))
+	for _, mergeKey := range mergeKeys {
+		deDuplicatedMergeKeys[mergeKey] = false
+	}
+
+	paramSetsByMergeKey := make(map[string]map[string]string, len(paramSets))
+	for _, paramSet := range paramSets {
+		paramSetKey := make(map[string]string)
+		for mergeKey := range deDuplicatedMergeKeys {
+			paramSetKey[mergeKey] = paramSet[mergeKey]
+		}
+		paramSetKeyJson, err := json.Marshal(paramSetKey)
+		if err != nil {
+			return nil, err
+		}
+		paramSetKeyString := string(paramSetKeyJson)
+		if _, exists := paramSetsByMergeKey[paramSetKeyString]; exists {
+			return nil, NonUniqueParamSets
```

Comment:
It couldn't hurt to report the specific `paramSetKeyString` that is causing the problem, in the error text, to help the user figure out where the problem is.